komorebi-bar/src/widgets/applications.rs

```diff
@@ -0,0 +1,326 @@
+            self.launch_if_ready();
+        }
+    }
+
+    /// Draws the application's icon within the UI if available,
+    /// or falls back to a default placeholder icon.
+    #[inline]
+    fn draw_icon(&self, ctx: &Context, ui: &mut Ui, icon_config: &IconConfig) {
+        if let Some(icon) = &self.icon {
+            icon.draw(ctx, ui, icon_config);
+        } else {
+            Icon::draw_fallback(ui, Vec2::splat(icon_config.size));
+        }
+    }
+
+    /// Displays the application's name as a non-selectable label within the UI.
+    #[inline]
+    fn draw_name(&self, ui: &mut Ui) {
+        ui.add(Label::new(&self.name).selectable(false));
+    }
+
+    /// Attempts to launch the specified command in a separate thread if enough time has passed
+    /// since the last launch. This prevents repeated launches from rapid consecutive clicks.
+    ///
+    /// Errors during launch are logged using the `tracing` crate.
+    pub fn launch_if_ready(&mut self) {
+        let now = Instant::now();
+        if now.duration_since(self.last_launch) < MIN_LAUNCH_INTERVAL {
+            return;
+        }
+
+        self.last_launch = now;
+        let command_string = self.command.clone();
+        // Launch the application in a separate thread to avoid blocking the UI
+        std::thread::spawn(move || {
+            if let Err(e) = Command::new("cmd").args(["/C", &command_string]).spawn() {
+                tracing::error!("Failed to launch command '{}': {}", command_string, e);
+            }
+        });
+    }
+}
+
+/// Holds decoded image data to be used as an icon in the UI.
+#[derive(Clone, Debug)]
+pub enum Icon {
+    /// RGBA image used for rendering the icon.
+    Image(RgbaImage),
+    /// Text-based icon, e.g. from a font like Nerd Fonts.
+    Text(String),
+}
+
+impl Icon {
+    /// Attempts to create an `Icon` from the given `AppConfig`.
+    /// Loads the image from a specified icon path or extracts it from the application's
+    /// executable if the command points to a valid executable file.
+    #[inline]
+    pub fn try_from(config: &AppConfig) -> Option<Self> {
+        if let Some(icon) = config.icon.as_deref().map(str::trim) {
+            if !icon.is_empty() {
+                let path = std::path::Path::new(icon);
+                if path.is_file() {
+                    match image::open(path).as_ref().map(DynamicImage::to_rgba8) {
+                        Ok(image) => return Some(Icon::Image(image)),
+                        Err(err) => {
+                            tracing::error!("Failed to load icon from {}, error: {}", icon, err)
+                        }
+                    }
+                } else {
+                    return Some(Icon::Text(icon.to_owned()));
+                }
+            }
+        }
+
+        let path = std::path::Path::new(&config.command);
+        if path.is_file() {
+            return windows_icons::get_icon_by_path(&config.command).map(Icon::Image);
```

Comment:
This is the bit where if this doesn't return an icon, we should return whatever we get from `windows_icons_fallback`, even if that also returns nothing

Reply:
Fixed